components/proxy/src/main/java/com/hotels/styx/proxy/backends/file/FileChangeMonitor.java

```diff
@@ -0,0 +1,155 @@
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * Monitors a file system object and notifies the consumer of any changes.
+ */
+public class FileChangeMonitor implements FileMonitor {
+    private final Path monitoredFile;
+    private final ScheduledExecutorService executor = newSingleThreadScheduledExecutor();
+    private final long pollPeriod;
+    private final TimeUnit timeUnit;
+    private ScheduledFuture<?> monitoredTask;
+
+    private AtomicReference<FileTime> lastChangedTime = new AtomicReference<>(FileTime.fromMillis(0));
+    private AtomicReference<HashCode> hashCode = new AtomicReference<>();
+    private volatile boolean performHashCheck;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FileChangeMonitor.class);
+
+    @VisibleForTesting
+    FileChangeMonitor(String monitoredFile, long pollPeriod, TimeUnit timeUnit) {
+        requireExists(requireNonNull(monitoredFile));
+        this.monitoredFile = Paths.get(monitoredFile);
+        this.pollPeriod = pollPeriod;
+        this.timeUnit = requireNonNull(timeUnit);
+        this.hashCode.set(HashCode.fromLong(0));
+    }
+
+    public FileChangeMonitor(String monitoredFile) {
+        this(monitoredFile, 1, SECONDS);
+    }
+
+    @Override
+    public void start(Listener listener) {
+        synchronized (this) {
+            if (monitoredTask != null) {
+                throw new IllegalStateException(format("File monitor for '{}' is already started", monitoredFile));
+            }
+
+            monitoredTask = executor.scheduleAtFixedRate(newPollerTask(listener), pollPeriod, pollPeriod, timeUnit);
+        }
+    }
+
+    private Runnable newPollerTask(Listener listener) {
+        return () -> {
+            if (!exists(monitoredFile)) {
+                LOGGER.debug("Monitored file does not exist. Path={}", monitoredFile);
+                return;
+            }
+
+            if (!isReadable(monitoredFile)) {
+                LOGGER.debug("Monitored file is no longer readable. Path={}", monitoredFile);
+                return;
+            }
+
+            if (modificationTimeChanged(monitoredFile)) {
+                hashCode.set(fileContentMd5(monitoredFile));
+                performHashCheck = true;
+
+                listener.fileChanged();
+                return;
+            }
+
+            if (performHashCheck && contentHashChanged(monitoredFile)) {
+                listener.fileChanged();
+            }
+        };
+    }
+
+    private boolean modificationTimeChanged(Path monitoredFile) {
+        try {
+            FileTime current = getLastModifiedTime(monitoredFile);
+            FileTime previous = lastChangedTime.getAndSet(current);
+            boolean changed = !previous.equals(current);
+            LOGGER.debug("modificationTimeChange probe. Changed={}", changed);
+            return changed;
+        } catch (IOException e) {
+            LOGGER.debug("modificationTimeChange probe failed. Cause={}", e);
```

Comment:
Why not add this information into the exception message?

Reply:
Hmm... perhaps you are correct.